pkg/controller/gkenetworkparamset/gkenetworkparamset_controller.go

```diff
@@ -333,37 +333,55 @@ func (c *Controller) syncNetworkWithGNP(ctx context.Context, network *networkv1.
 	return nil
 }
 
-func (c *Controller) handleGKENetworkParamSetDelete(ctx context.Context, params *networkv1.GKENetworkParamSet) error {
+// handleGNPDelete checks to see if its safe to delete the GNP resource before calling executeGNPDelete on it
+func (c *Controller) handleGNPDelete(ctx context.Context, params *networkv1.GKENetworkParamSet) error {
 	if params.Status.NetworkName == "" {
-		removeFinalizerInPlace(params)
-		return nil
+		return c.executeGNPDelete(ctx, params, nil)
 	}
 
 	network, err := c.networkClientset.NetworkingV1().Networks().Get(ctx, params.Status.NetworkName, v1.GetOptions{})
 	if err != nil {
 		if errors.IsNotFound(err) {
-			removeFinalizerInPlace(params)
-			return nil
+			return c.executeGNPDelete(ctx, params, nil)
 		}
 		return err
 	}
 
 	networkStillRefersToGNP := network.Spec.ParametersRef != nil && strings.EqualFold(network.Spec.ParametersRef.Kind, gnpKind) && network.Spec.ParametersRef.Name == params.Name
 	if !networkStillRefersToGNP {
-		removeFinalizerInPlace(params)
-		return nil
+		return c.executeGNPDelete(ctx, params, network)
 	}
 
 	if networkStillRefersToGNP && !network.InUse() {
-		removeFinalizerInPlace(params)
-		return nil
+		return c.executeGNPDelete(ctx, params, network)
 	}
 
 	// if the network is in use, this GNP object will get reconciled again when the network's in use status changes.
 
 	return nil
 }
 
+func (c *Controller) executeGNPDelete(ctx context.Context, params *networkv1.GKENetworkParamSet, network *networkv1.Network) error {
+	removeFinalizerInPlace(params)
+
+	if network != nil {
+		newNetwork := network.DeepCopy()
```

Comment:
You can skip the DeepCopy. This isn't very cheap. You can directly update the status.

Reply:
Are you sure? I thought we couldn't modify objects directly because of cache issues. However it might not matter in this case.

Reply:
I am not sure. I defer to your judgement.

Reply:
I'm going to keep the DeepCopy.

Reply:
if the `networkReference`the cache you can not mutate the object, since is a reference and all the controllers using the same cache will have the object mutated, see

https://github.com/kubernetes/kubernetes/issues/116694
https://github.com/kubernetes/kubernetes/issues/89157